pkg/certificate/manager.go

```diff
@@ -63,29 +66,56 @@ func (m *Manager) GetTrustDomain() string {
 	return ""
 }
 
+// ShouldRotate determines whether a certificate should be rotated.
+func (m *Manager) shouldRotate(c *Certificate) bool {
+	// The certificate is going to expire at a timestamp T
+	// We want to renew earlier. How much earlier is defined in renewBeforeCertExpires.
+	// We add a few seconds noise to the early renew period so that certificates that may have been
+	// created at the same time are not renewed at the exact same time.
+	intNoise := rand.Intn(noiseSeconds) // #nosec G404
+	secondsNoise := time.Duration(intNoise) * time.Second
+	renewBefore := RenewBeforeCertExpires + secondsNoise
+	if time.Until(c.GetExpiration()) <= renewBefore {
+		log.Info().Msgf("Cert %s should be rotated; expires in %+v; renewBefore is %+v",
+			c.GetCommonName(),
+			time.Until(c.GetExpiration()),
+			renewBefore)
+		return true
+	}
+
+	m.mu.RLock()
+	validatingIssuer := m.validatingIssuer
+	signingIssuer := m.signingIssuer
+	m.mu.RUnlock()
```

Comment:
not quite, since these are addresses, and we are changing the address, not the content of the address, the old address (which is different than a newly changed address) is still valid.

See https://go.dev/play/p/QIW1jMcor2j

Reply:
i should clarify here, this may indeed grab an old issuer, and begin signing with that, which is ok, since we'll achieve eventual consistency

the lock is strictly to prevent data races with access to the field itself, not to guarantee that the issuer is still current

Reply:
@nojnhuh what are your thoughts on this?

This is how I understand it. Once we create an issuer, the issuer itself should never be modified. We will only ever create a new issuer and update the pointer to the address of that new issuer. So while we aren't doing a deep copy of the issuer, it is guaranteed that the issuer at that address will never be updated. Therefore we don't need locks for reads. We do need locks when we update the validating and signing issuer pointers. @steeling is that on track?